src/disco/shred/fd_fec_resolver.h

```diff
@@ -0,0 +1,162 @@
+                 |                      |--- if spilled ----    |
+                 ------------------------                  |    |
+                       ^                                   |    |
+                       |                                   V    |
+                   --------                                |    |
+                   |      |   -                            |    V
+                   | Free |    \                           |    |
+                   |      |     >=partial_depth            |    |
+                   | FEC  |     /                          |    |
+                   | sets |    |                           |    |
+                   |      |   _                            |    |
+                   --------                                V    |
+                      ^  ^                                 |    |
+                      |  |------<---------------<----------|    |
+                      |                                         |
+                   --------                                     |
+                   |      | -                                   |
+                   | Comp |  \                                  |
+                   |leted |   complete_depth                    |
+                   |      |  /                                  V
+                   | FEC  |  |                                  |
+                   | sets |  |                                  |
+                   |      |  -                                  |
+                   --------                                     |
+                      ^                                         |
+                      |-------------------<----------------------
+
+   When a shred arrives for a new FEC set, we pull an entry from the
+   head of the free FEC set queue.  If that would result in more than
+   depth sets in progress, we spill the oldest in progress set, and
+   insert it at the tail of the free set queue.  When we complete a set,
+   we remove it from the in progress set, add it to the tail of the
+   completed FEC set queue, and move one element from the head of the
+   completed queue to the tail of the free queue.
+
+   Since the completed queue only advances when an FEC set is completed,
+   any complete FEC set stays in that queue for at least complete_depth
+   completions.
+
+   It might seems like this system is overkill and one combined longer
+   freelist would suffice, but that's incorrect. Consider the following
+   scenario: Suppose we've just completed FEC set A.  Then we receive
+   lots of shreds for lots of FEC sets, but never enough to complete any
+   one of them.  We churn through the whole singular freelist with
+   spilling until A is at the head of the freelist.  We receive shreds
+   for another FEC set, so we reuse the memory we used for A, and it's
+   the first set that we can complete, so we end up returning the same
```

Comment:
I don't understand this part. When we "reuse the memory we used for A", it would be considered dirty for A, so the downstream consumer wouldn't use it anymore, then if the "new A" completes, it goes onto the freelist? What does it mean to "return the same memory twice in a row"

Reply:
Is this more clear?